### ISSUE TITLE
Smarter Auto Increment

### ISSUE THREAD

**Issue by jamboNum5** [grokability/snipe-it#10537]
### Snipe-IT Version

Version v5.3.0 - build 6470 (master)

### Operating System

CentOS

### Web Server

Apache

### PHP Version

7

### Is your feature request related to a problem? Please describe. A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]

Our asset system can be pretty busy sometimes, we seem to end up with people trying to add assets at the same time and getting the classic message stating that the Asset Tag needs to be unique. The person adding the asset then has to manually increment the number to allow the asset to get created in the system.

What seems to happen is every now and then, someone will increment by two rather than one which throw the system count out completely. This obviously becomes a vicious cycle which breaks the new asset counter.

In the recent version I updated to, this was compounded. When the unique asset number is deemed invalid, the form resets the information in the fields and means more manual entry for people.

I can reset the number in the settings, but this happens so often it seems it could be automated to fix itself. 

### Describe the solution you'd like A clear and concise description of what you want to happen.

I'm hoping that this can be sorted through some things like the following options:

- An AJAX check can happen when the submit button is press to check if the asset tag number is still ok to use, auto incrementing if it isn't.
- Auto generated tag number is created after new asset form is submitted to avoid the situation where two people open a webpage around the same time.

### Describe alternatives you've considered A clear and concise description of any alternative solutions or features you've considered.

Ever so often I reset the main number, but the system is so busy that people end up incrementing too much and it's pretty impossible to police their actions.

**Comment by jamboNum5** [comment]
Hi Brady

Do you think this feature is something this would be possible?

Thanks - J

**Comment by uberbrady** [comment]
Yes, I think this is possible. The most "correct" way to do it is to not show the asset tag on the create asset screen at all (maybe with some instruction like "leave blank to auto-generate"), and only show it after the asset has been actually created. But I suspect that won't work with a lot of people's workflows.

Otherwise, we could let the code handle the incrementing for you (*if* you have asset-tag auto-generation available) on failed submit - maybe with a modified error message saying "we incremented it for you, hit submit again" or something like that, allowing the user to submit it again.

As I'm describing it I'm actually liking option #1 better, but I will defer to @snipe on how she thinks this is best handled.

**Comment by snipe** [comment]
They're unfortunately both awful :( We've considered ways around this for a while , but at the end of the day, it's going to impact some percentage of people's workflows adversely :( If we try to be too clever (alert saying "I know we told you it was gonna be this asset tag, but now it's this one - but your asset was saved") or using literal web sockets or polling to live update, there are going to be folks who are filling in some other form or software with that they believe to be true at the time of page load. If we remove the asset tag from that screen if you have incrementing enabled, people will complain that now they don't have the info they need to enter into XYZ other software or HR form or what have you.

Best compromise I can think of is to have a note under the asset tag box that says "You have auto-incrementing enabled - this is *probably* going to be your asset tag, but hit save to find out!"

I hate it, but I can't think of anything less frictiony

**Comment by jamboNum5** [comment]
Thank you both for taking the your time to reply, I imagine this is a bit of a frustrating Bugbear 🐻

All the best, I'll set this to closed.

**Comment by uberbrady** [comment]
No, I think it's a valid concern, I'd like to leave it open, and I think I have an idea for how I want to do it:

First off, **if** you have auto-increment asset tags turned on, I'd like the asset tag field to be left blank, with "placeholder text" (the light gray text) saying "Automatically generated" or something like that (yay, another translation string! :( )

Then, once you create an asset, on the "Asset successfully created" 'flash' message, I'd like it to say something like: "Asset succesfully created. Generated asset tag is: BLAH00007" or something like that. (Yay! Another _other_ translation string, now with a parameter! :( )

I think this will satisfy the bulk of the use cases, and with some server-side transactional logic we can generate incrementing asset tags in a way that actually makes sense for our use cases - maybe with a bit of a loop and some retry logic? (With some kind of sanity-check "I tried to pick an asset tag for this asset 100 times and couldn't" kind of emergency back-off. I'm fine with not translating that string, as it ought not to happen...but maybe if you set your 'next autoincrement number' to like 1 and you already have 1000 assets with asset tags starting at 1, then it might show up. But that happens already, right now, so maybe that's not so bad....)

I don't think I have *all* the answers for this, but I think it's something we could at least do a little bit better.

**Comment by jamboNum5** [comment]
This sounds a great, I can't say I'm any good with coding but if there is any testing I can help with please let me know. Many thanks Brady

**Comment by jamboNum5** [comment]
As the ID tag was getting incorrectly incremented by users and breaking the numbering system, I have just added some CSS to remove the ability to override this input field. (If you don't know how to do this, the CSS is set in the branding section of the settings).

```
input#asset_tag {
    pointer-events: none;
}
```
